pkg/rc/replication_controller.go

```diff
@@ -625,81 +663,280 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
 	if err != nil {
-		return util.Errorf("Could not commit CASTxn: %s", err)
+		return "", "", util.Errorf("Could not commit CASTxn: %s", err)
+	}
+
+	return newNode, oldNode, nil
+}
+
+func (rc *replicationController) scheduleWithSession(newNode types.NodeName) error {
+	rc.logger.NoFields().Infof("Scheduling with session on %s", newNode)
+	rc.mu.Lock()
+	manifest := rc.Manifest
+	rc.mu.Unlock()
+
+	manifestBytes, err := manifest.Marshal()
+	if err != nil {
+		return err
+	}
+
+	key, err := consul.PodPath(consul.INTENT_TREE, newNode, manifest.ID())
+	if err != nil {
+		return err
 	}
 
+	name := fmt.Sprintf("wait-on-%s-health-%s", newNode, rc.ID())
+
+	ctx, ctxCancel := context.WithCancel(context.Background())
+	defer func() {
+		// We close the context on error to prevent a leak but leave it
+		// otherwise so it can be used to cancel the session. We write the
+		// defer this way so that reassignments to err are noticed and the
+		// final value is used.
+		if err != nil {
+			ctxCancel()
+		}
+	}()
+
+	// By passing ctx below, ctxCancel (defined above) can be used to destroy
+	// the session when the node transfer finishes or is stopped
+	_, session, err := consul.SessionContext(ctx, rc.consulClient, name)
+	if err != nil {
+		return err
+	}
+
+	writeCtx, writeCancel := transaction.New(context.Background())
+	defer writeCancel()
+
+	txnUnlocker, err := session.LockIfKeyNotExistsTxn(writeCtx, key, manifestBytes)
+	if err != nil {
+		return err
+	}
+
+	ok, resp, err := transaction.Commit(writeCtx, rc.txner)
+	switch {
+	case err != nil:
+		return err
+	case !ok:
+		// assign to err instead of just returning it so that deferred func sees
+		// the assignment
+		err = util.Errorf("transaction violation scheduling new node: %s", transaction.TxnErrorsToString(resp.Errors))
+		return err
+	}
+
+	rc.nodeTransfer.sessionCancel = ctxCancel
+	rc.nodeTransfer.unlocker = txnUnlocker
+
 	return nil
 }
 
-func (rc *replicationController) transferPetNodes(ineligibleCurrent []types.NodeName) error {
-	errMsg := fmt.Sprintf("RC has scheduled %d ineligible nodes: %s", len(ineligibleCurrent), ineligibleCurrent)
-	err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+func (rc *replicationController) doBackgroundNodeTransfer() {
+	rc.isWatchingHealth = true
+	ok := rc.watchHealth()
+	if ok {
+		err := rc.finishTransfer()
+		if err != nil {
+			rc.logger.WithError(err).Errorln("could not do final node transfer transaction")
+		}
+	}
+
+	rc.isWatchingHealth = false
+
+	rc.nodeTransfer.sessionCancel()
+
+	rc.nodeTransfer = nodeTransfer{}
+	err := rc.rcStatusStore.Delete(rc.ID())
+	if err != nil {
+		rc.logger.WithError(err).Errorln("error deleting transfer status during cleanup")
+	}
+}
+
+func (rc *replicationController) watchHealth() bool {
+	rc.mu.Lock()
+	podID := rc.Manifest.ID()
+	rc.mu.Unlock()
+
+	resultCh := make(chan health.Result)
+	errCh := make(chan error)
+	healthQuitCh := make(chan struct{})
+	defer func() {
+		close(resultCh)
```

Comment:
unfortunately i think closing this isn't safe, because the way `WatchNodeService` is written it might try to send on this channel after it is closed.

I hate to give you more work for this PR, but probably `WatchNodeService()` should create `resultCh` and `errCh` itself, return them, and then take care of closing them after it sees `healthQuitCh` close.

Then you would call it like this:

```
defer close(healthQuitCh)
resultCh, errCh := rc.healthChecker.WatchNodeService(healthQuitCh)
```
which would guarantee that we have a clean shutdown.

Reply:
I'll open up a new PR refactoring `WatchNodeService` to use `WatchSingle` and to avoid the panic case